### ISSUE TITLE
passport.socketio doesn't work with passport 0.3.0:  TypeError: Cannot set property 'user' of undefined

### ISSUE THREAD

**Issue by garretraziel** [jfromaniello/passport.socketio#109]
Current version from npm (3.5.1) doesn't work with newest passport (0.3.0). Trying to log in with redirect to page with socket.io connection results in:

```
TypeError: Cannot set property 'user' of undefined
   at /home/.../node_modules/passport.socketio/node_modules/passport/lib/http/request.js:50:35
   at pass (/home/.../node_modules/passport/lib/authenticator.js:267:43)
   at serialized (/home/.../node_modules/passport/lib/authenticator.js:276:7)
   at serializeUser (/home/.../lib/secure.js:48:5)
   at pass (/home/.../node_modules/passport/lib/authenticator.js:284:9)
   at Authenticator.serializeUser (/home/.../node_modules/passport/lib/authenticator.js:289:5)
   at IncomingMessage.req.login.req.logIn (/home/.../node_modules/passport.socketio/node_modules/passport/lib/http/request.js:48:29)
   at Strategy.strategy.success (/home/.../node_modules/passport/lib/middleware/authenticate.js:228:13)
   at verified (/home/.../node_modules/passport-local/lib/strategy.js:83:10)
   at /home/.../lib/secure.js:33:32
```

Downgrading passport to 0.2.2 solves this problem. It looks like the same problem as here: jaredhanson/passport#400

**Comment by busheezy** [comment]
I have been going damn near insane because of this.  I have literally toyed around with this for four days and didn't see this issue when I was here last.

**Comment by jfromaniello** [comment]
I don't see any file from passport.socketio in your stacktrace. 

I am not sure why you are using the the passport module inside passport.socketion though.

**Comment by garretraziel** [comment]
I think that it may be caused by the fact that latest version of passport.socketio in npm isn't 3.6.0 but 3.5.1. Version 3.5.1 doesn't require passport 0.3.0 but 0.2.0. In Passport 0.3.0, it was changed how `req.session.passport` is created. So I log in using Passport from my project dependencies (0.3.0), but then session is accessed from Passport 0.2.0 (from passport.socketio dependencies) and it fails with this error.

In other words, my project uses Passport 0.3.0, but passport.socketio uses Passport 0.2.0 (it has it as its dependency), session is created using Passport 0.3.0, but as soon as I use passport.socketio, Passport 0.2.0 access this session and it fails.

**Comment by jfromaniello** [comment]
@garretraziel yes, try initializing passport.socketio as follows:

```js
io.set('authorization', passportSocketIo.authorize({
  passport: require('passport'),
  cookieParser: express.cookiePars..
```

**Comment by busheezy** [comment]
This fixed my problem.  Thank you for your help and work.

**Comment by garretraziel** [comment]
Ok, this seems to fix problem for me also.

**Comment by jfromaniello** [comment]
Glad to help guys!

This has caused issues in the past, maybe I should remove the passport dependency and always ask the user to provide an instance of passport.

**Comment by garretraziel** [comment]
@jfromaniello I think that this is a good idea - you will not run into incompatibilities between several versions of passport.

**Comment by busheezy** [comment]
I thought I was fine but I'm still having errors on my test deployment server.

```
TypeError: Cannot set property 'user' of undefined
    at /home/foshata/site/foshata/node_modules/passport.socketio/node_modules/passport/lib/http/request.js:50:35
    at pass (/home/foshata/site/foshata/node_modules/passport/lib/authenticator.js:267:43)
    at serialized (/home/foshata/site/foshata/node_modules/passport/lib/authenticator.js:276:7)
    at /home/foshata/site/foshata/app.js:104:3
    at pass (/home/foshata/site/foshata/node_modules/passport/lib/authenticator.js:284:9)
    at Authenticator.serializeUser (/home/foshata/site/foshata/node_modules/passport/lib/authenticator.js:289:5)
    at IncomingMessage.req.login.req.logIn (/home/foshata/site/foshata/node_modules/passport.socketio/node_modules/passport/lib/http/request.js:48:29)
    at Strategy.strategy.success (/home/foshata/site/foshata/node_modules/passport/lib/middleware/authenticate.js:228:13)
    at verified (/home/foshata/site/foshata/node_modules/passport-local/lib/strategy.js:83:10)
    at [object Object].<anonymous> (/home/foshata/site/foshata/app.js:69:14)
    at [object Object].tryCatcher (/home/foshata/site/foshata/node_modules/sequelize/node_modules/bluebird/js/main/util.js:26:23)
    at Promise._settlePromiseFromHandler (/home/foshata/site/foshata/node_modules/sequelize/node_modules/bluebird/js/main/promise.js:503:31)
    at Promise._settlePromiseAt (/home/foshata/site/foshata/node_modules/sequelize/node_modules/bluebird/js/main/promise.js:577:18)
    at Promise._settlePromises (/home/foshata/site/foshata/node_modules/sequelize/node_modules/bluebird/js/main/promise.js:693:14)
    at Async._drainQueue (/home/foshata/site/foshata/node_modules/sequelize/node_modules/bluebird/js/main/async.js:123:16)
    at Async._drainQueues (/home/foshata/site/foshata/node_modules/sequelize/node_modules/bluebird/js/main/async.js:133:10)
```

**Comment by jfromaniello** [comment]
Given this line in the stacktrace: 

```node_modules/passport.socketio/node_modules/passport```

It seems you are not passing your passport object. Can you show me how you config this lib?

**Comment by busheezy** [comment]
I must be losing my mind.  I came back to it and it is working.  I must have been doing something silly.  Sorry to bother.  Thanks for your help.

**Comment by busheezy** [comment]
I am 99 percent sure I have changed nothing and it is no longer working.

I've been using babel and I'm wondering if it is doing something I don't intend.

**Comment by busheezy** [comment]
I'm currently using a npm-shrinkwrap to force it.  I'm not sure why I can't pass passport into it.

I'll look into it more in the morning.

**Comment by buunguyen** [comment]
@jfromaniello instead of removing the `passport` dependency altogether, maybe you should consider declaring it as a [peer dependency](https://docs.npmjs.com/files/package.json#peerdependencies)?

**Comment by garretraziel** [comment]
@busheezy Are you sure that you are using correct passport? According to this:

    /home/foshata/site/foshata/node_modules/passport.socketio/node_modules/passport/lib/http/request.js:50:35

it looks like it's not using version you have imported, but rather its own version.

**Comment by mwager** [comment]
Experiencing the same issue since upgrading my project dependency on passport to 0.3.0. Passing my passport instance and installing passport.socketio from github fixes it. 

Why you are not doing npm publish?

**Comment by jfromaniello** [comment]
as suggested by @buunguyen , I moved passport to [peerDependencies](https://docs.npmjs.com/files/package.json#peerdependencies).

Please notice that the behaviour is different from npm v2 to npm v3. Npm 2 installs the dependency if is missing from your own package.json, while npm 3 will show a warning but it will never install the dependency.

I hope this solve most of the issues.

I published this change as `passport.socketio@3.6.1`.

**Comment by mwager** [comment]
thx!

**Comment by busheezy** [comment]
Thanks a ton for your work and help.

**Comment by scottie** [comment]
omg finally what a head screw that was!! thanks
